alvr/sockets/src/stream_socket/mod.rs

```diff
@@ -0,0 +1,398 @@
+// transmission of a big packet before. If we allow interleaving shards, small packets can be
+// transmitted quicker, with only minimal latency increase for the ongoing transmission of the big
+// packet.
+// Note: We can't clone the underlying socket for each StreamSender and the mutex around the socket
+// cannot be removed. This is because we need to make sure at least shards are written whole.
+
+mod tcp;
+mod udp;
+
+use alvr_common::{
+    AnyhowToCon, ConResult, HandleTryAgain, ToCon, anyhow::Result, parking_lot::Mutex,
+};
+use alvr_session::{DscpTos, SocketBufferSize, SocketProtocol};
+use bincode::config;
+use serde::{Serialize, de::DeserializeOwned};
+use std::{
+    cmp::Ordering,
+    collections::HashMap,
+    marker::PhantomData,
+    mem,
+    net::{IpAddr, TcpListener, UdpSocket},
+    ops::Range,
+    sync::{Arc, mpsc},
+    time::Duration,
+};
+
+trait MultiplexedSocketWriter {
+    // Note: consts are not trait-safe, we require a method
+    fn payload_offset(&self) -> usize;
+
+    fn send(&mut self, stream_id: u16, packet_index: u32, buffer: &mut Vec<u8>) -> Result<()>;
+}
+
+struct ReconstructedPacket {
+    index: u32,
+    buffer: Vec<u8>,
+}
+
+struct StreamRecvQueues {
+    used_buffer_sender: mpsc::Sender<Vec<u8>>,
+    used_buffer_receiver: mpsc::Receiver<Vec<u8>>,
+    packet_queue: mpsc::Sender<ReconstructedPacket>,
+}
+
+trait MultiplexedSocketReader {
+    fn payload_offset(&self) -> usize;
+
+    fn recv(&mut self, stream_queues: &HashMap<u16, StreamRecvQueues>) -> ConResult;
+}
+
+/// Memory buffer that contains a hidden prefix
+#[derive(Default)]
+pub struct Buffer<H = ()> {
+    inner: Vec<u8>,
+    raw_payload_offset: usize, // this corresponds to prefix + header
+    _phantom: PhantomData<H>,
+}
+
+impl<H> Buffer<H> {
+    /// Length of payload (without prefix)
+    #[must_use]
+    pub fn len(&self) -> usize {
+        self.inner.len() - self.raw_payload_offset
+    }
+
+    #[must_use]
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Get the whole payload of the buffer
+    #[must_use]
+    pub fn get(&self) -> &[u8] {
+        &self.inner[self.raw_payload_offset..]
+    }
+
+    /// The buffer starts with no raw payload allocated space. If the range is outside the valid
+    /// range, new space will be allocated
+    /// NB: the offset parameter is applied on top of the internal offset of the buffer
+    #[must_use]
+    pub fn get_range_mut(&mut self, range: Range<usize>) -> &mut [u8] {
+        let required_size = self.raw_payload_offset + range.end;
+        if required_size > self.inner.len() {
+            self.inner.resize(required_size, 0);
+        }
+
+        &mut self.inner[self.raw_payload_offset..][range]
+    }
+
+    /// If length > current length, allocate more space
+    pub fn set_len(&mut self, length: usize) {
+        self.inner.resize(self.raw_payload_offset + length, 0);
+    }
```

Comment:
I don't think this comment is needed, or else it, because resizing shouldn't allocate if there's already enough capacity (and even if that's not intended "allocating space" seems like "allocating memory"), sth like:
> Behaves like `std::vec::resize` in respect to the payload.